clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -109,9 +109,9 @@ public Metrics(MetricConfig defaultConfig, List<MetricsReporter> reporters, Time
      */
     public Metrics(MetricConfig defaultConfig, List<MetricsReporter> reporters, Time time, boolean enableExpiration) {
         this.config = defaultConfig;
-        this.sensors = new CopyOnWriteMap<>();
-        this.metrics = new CopyOnWriteMap<>();
-        this.childrenSensors = new CopyOnWriteMap<>();
+        this.sensors = new ConcurrentHashMap<>();
```

Comment:
I added some numbers in the pull description. Copying them here with some more data. Are you looking for anything more specific? I tested it to work upto 500k sensors which is much higher than we can add now.

ConcurrentHashMap Time to create :
100k sensors (1.5 seconds)
200k sensors (3 seconds)
400k sensors (9 seconds)
500k sensors (14 seconds)

TreeMap
100k sensors (2 seconds)
200k sensors (4 seconds)
400k sensors (10 seconds)
500k sensors (14 seconds)